librtt/librtt.c

```diff
@@ -0,0 +1,432 @@
+			dstBuf[wr] = *srcBuf++;
+			wr = (wr + 1) & sz;
+			todo--;
+		}
+
+		rtt->channel[ch].rd = wr;
+	}
+
+	performCacheOp(librtt_common.cleanFn, (void *)dstBuf, sz + 1, rd, wr);
+	rtt->channel[ch].wr = wr;
+
+	dataMemoryBarrier();
+
+	return count - todo;
+}
+
+
+ssize_t librtt_rxAvail(unsigned int ch)
+{
+	volatile struct rtt_desc *rtt = librtt_common.rtt;
+	if (librtt_checkRx(ch) < 0) {
+		return 0;
+	}
+
+	dataMemoryBarrier();
+	ch += librtt_common.rtt->txChannels;
+	unsigned int sz = rtt->channel[ch].sz - 1;
+	unsigned int rd = rtt->channel[ch].rd & sz;
+	unsigned int wr = rtt->channel[ch].wr & sz;
+
+	if (rd > wr) {
+		return sz + 1 - (rd - wr);
+	}
+	else {
+		return wr - rd;
+	}
+}
+
+
+ssize_t librtt_txAvail(unsigned int ch)
+{
+	volatile struct rtt_desc *rtt = librtt_common.rtt;
+	if (librtt_checkTx(ch) < 0) {
+		return 0;
+	}
+
+	dataMemoryBarrier();
+	unsigned int sz = rtt->channel[ch].sz - 1;
+	unsigned int rd = (rtt->channel[ch].rd + sz) & sz;
+	unsigned int wr = rtt->channel[ch].wr & sz;
+
+	if (wr > rd) {
+		return sz + 1 - (wr - rd);
+	}
+	else {
+		return rd - wr;
+	}
+}
+
+
+void librtt_rxReset(unsigned int ch)
+{
+	dataMemoryBarrier();
+	if (librtt_checkRx(ch) < 0) {
+		return;
+	}
+
+	ch += librtt_common.rtt->txChannels;
+	librtt_common.rtt->channel[ch].rd = librtt_common.rtt->channel[ch].wr;
+	dataMemoryBarrier();
+}
+
+
+void librtt_txReset(unsigned int ch)
+{
+	dataMemoryBarrier();
+	if (librtt_checkTx(ch) < 0) {
+		return;
+	}
+
+	librtt_common.rtt->channel[ch].wr = librtt_common.rtt->channel[ch].rd;
+	dataMemoryBarrier();
+}
+
+
+int librtt_txCheckReaderAttached(unsigned int ch)
+{
+	if (librtt_checkTx(ch) < 0) {
+		return 0;
+	}
+
+	unsigned int rd = librtt_common.rtt->channel[ch].rd;
+	if (librtt_common.lastRd[ch] != rd) {
+		librtt_common.lastRd[ch] = rd;
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static int librtt_verifyChannel(volatile struct rtt_desc *rtt, unsigned int ch, unsigned char *buffers, size_t buffersSize)
+{
+	unsigned char *bufEnd = buffers + buffersSize;
+	size_t sz = rtt->channel[ch].sz;
+	/* Check buffer size is non-zero and power of 2 */
+	if ((sz == 0) || ((sz & (sz - 1)) != 0)) {
+		return -EINVAL;
+	}
+
+	if ((rtt->channel[ch].ptr < buffers) || (rtt->channel[ch].ptr >= bufEnd)) {
+		return -EINVAL;
+	}
+
+	if (((rtt->channel[ch].ptr + rtt->channel[ch].sz) <= buffers) || ((rtt->channel[ch].ptr + rtt->channel[ch].sz) > bufEnd)) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+
+int librtt_verify(void *addr, size_t cbSize, void *buffers, size_t buffersSize, librtt_cacheOp_t invalFn, librtt_cacheOp_t cleanFn)
+{
+	int n, ret;
+	if ((cbSize < sizeof(struct rtt_desc))) {
+		return -EINVAL;
+	}
+
+	volatile struct rtt_desc *rtt = addr;
+	for (n = 0; n < librtt_tagLength; n++) {
+		if (librtt_tagReversed[n] != rtt->tag[librtt_tagLength - 1 - n]) {
+			break;
+		}
+	}
+
+	if (n != librtt_tagLength) {
+		return -EINVAL;
+	}
+
+	if ((rtt->txChannels > LIBRTT_TXCHANNELS) || (rtt->rxChannels > LIBRTT_RXCHANNELS)) {
```

Comment:
This is tricky. We should not rely on CB being what we want it to be - if plo/kernel need additional channels, then we could reuse just the lower ones (as limited by `LIBRTT_TXCHANNELS` and `LIBRTT_TXCHANNELS`) instead of rejecting them.

Reply:
Fixed, now the control blocks can have more channels than the limit. The channels beyond the limit won't be usable and won't be modified.